lightning/src/util/events.rs

```diff
@@ -82,6 +82,34 @@ impl_writeable_tlv_based_enum!(PaymentPurpose,
 	(2, SpontaneousPayment)
 );
 
+/// When the payment failure took place and extra details about it. [`Failure::OnPath`] may contain
+/// a [`NetworkUpdate`] that needs to be applied to the [`NetworkGraph`].
+///
+/// [`NetworkUpdate`]: crate::routing::gossip::NetworkUpdate
+/// [`NetworkGraph`]: crate::routing::gossip::NetworkGraph
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Failure {
+	/// We failed to initially send the payment and no HTLC was committed to. Contains a relevant
+	/// error message.
+	InitialSend(String),
```

Comment:
It doesn't implement `writeable` at the moment, though that wouldn't be too hard to add.

Reply:
FYI, I think we'll need to change the `APIError::InvalidRoute` to contain a `String`. I don't think we can implement `Readable` for a `&'static str` because it references data owned by the `read` function.